### PR TITLE
you-get: init at 0.4.390

### PR DESCRIPTION
**Pull request by ryneeverett**
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -23203,6 +23203,28 @@ in modules // {
     };
   };
 
+  you-get = buildPythonPackage rec {
```
Comment (joachifm) [inline pkgs/top-level/python-packages.nix]: `buildPythonApplication` is a convenient alternative; it's the same thing except it sets `namePrefix = ""` for you ...

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -23203,6 +23203,28 @@ in modules // {
...
+    # probably aren't suitable for a build environment anyway.
+    doCheck = false;
+
+    src = pkgs.fetchurl {
+      url = "mirror://pypi/y/you-get/${name}.tar.gz";
+      sha256 = "17hs0g9yvgvkmr7p1cz39mbbvb40q65qkc31j3ixc2f873gahagw";
+    };
+
+    meta = {
+      description = "a tiny command line utility to download media contents from the web";
```
Comment (joachifm) [inline pkgs/top-level/python-packages.nix]: Please capitalize the description, per the guidelines. Also, I suggest that you add yourself as the maintainer; adding orphaned packages doesn't make too much sense.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -23203,6 +23203,28 @@ in modules // {
...
+    src = pkgs.fetchurl {
+      url = "mirror://pypi/y/you-get/${name}.tar.gz";
+      sha256 = "17hs0g9yvgvkmr7p1cz39mbbvb40q65qkc31j3ixc2f873gahagw";
+    };
+
+    meta = {
+      description = "A tiny command line utility to download media contents from the web";
+      homepage = https://you-get.org;
+      license = licenses.mit;
+      maintainers = with maintainers; [ ryneeverett ];
```
Comment (joachifm) [inline pkgs/top-level/python-packages.nix]: Also add an entry to https://github.com/NixOS/nixpkgs/blob/master/lib/maintainers.nix

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -23203,6 +23203,28 @@ in modules // {
     };
   };
 
+  you-get = buildPythonApplication rec {
+    version = "0.4.390";
+    name = "you-get-${version}";
+    disabled = !isPy3k;
```
Comment (ryneeverett) [inline pkgs/top-level/python-packages.nix]: Should I remove this line since `doCheck` is false anyway?

**Comment by joachifm** [comment]
This package has an optional run-time dependency on ffmpeg. If you wanted to ensure that it's present, you could generate a wrapper for the you-get executable. I feel that as long as the app can resolve the dependency from PATH, if the user has installed it, then a wrapper is unnecessary --- crashing due to unmet optional dependencies is okay as far as I'm concerned, but you may feel differently.

**Comment by ryneeverett** [comment]
I'm fine with crashing if ffmpeg isn't installed, but isn't there a way I should express the optional dependency? Something like:

    propagatedBuildInputs = [ ] ++ optionals streaming [ ffmpeg ];

**Comment by joachifm** [comment]
`propagatedBuildInputs` means that if `X` depends on `Y`, then any dependency on `X` also pulls in `Y`. It's not a good mechanism for optional run-time dependencies.

**Comment by joachifm** [comment]
Can you check that the executable does not hard-code `/usr/bin`? If it does, it may need patching.

**Comment by ryneeverett** [comment]
It does not.

**Comment by joachifm** [comment]
Okay, that's good news, then installing ffmpeg into your environment should be sufficient. If people complain, you can always add in a wrapper later.

**Comment by joachifm** [comment]
Ah, I noticed that `meta.platforms` is missing. Please add it. Then, I think this is good to go.

**Comment by ryneeverett** [comment]
> Okay, that's good news, then installing ffmpeg into your environment should be sufficient. If people complain, you can always add in a wrapper later.

If there were a simple way to add optional dependencies to the metadata I would do that, but it doesn't sound like there is so that strategy sounds right. Out of curiosity though, is there a good example of implementing a "wrapper" you can point to?

**Comment by joachifm** [comment]
The general idea is to create a small shell script that sets up PATH and then execs the original executable. There's a helper named `wrapProgram` that does that for you, in fact I'd expect the `you-get` executable in this package to actually be a wrapper, generated by `buildPythonPackage`. You can also in many cases patch in the full path to the dependency, obviating the need to create a wrapper (e.g., replacing occurrences of `system("ffmpeg")` with `system("${ffmpeg}/bin/ffmpeg")`.

**Comment by ryneeverett** [comment]
Thanks for all the help!

**Pull request merged**